### PR TITLE
Track metadata: Replace TrackRecord and beat grid at once

### PR DESCRIPTION
**Pull request by uklotzde**
The last refactoring PR before fixing the state management in `DlgTrackInfo`. It is now possible to replace all track properties encapsulated in `TrackRecord` and the beat grid at once.

### CODE REVIEW

**Comment by Holzhaus** [comment]
Just wanted to say that I appreciate that you split changes from #2282 and #2656 into bite-sized chunks and open PRs separately, so that we don't end up with huge, unreviewable PR that stays unmerged forever 👍

**Comment by uklotzde** [comment]
> Just wanted to say that I appreciate that you went the extra step and split changes from #2282 and #2656 into bite-sized chunks, so that we don't end up with a huge, unreviewable PR that stays unmerged forever +1

I was even concerned that opening so many small PRs might be distracting. On the other hand it helps to focus on a single aspect and allows to review the PR if you are not familiar with the code.

While reviewing some smaller PRs myself recently I noticed how much more rewarding this is. Both for the submitter and for the reviewer.

**Comment by uklotzde** [comment]
Wait, maybe there is some redundancy now. I have to check again.

**Review by Holzhaus — COMMENTED** [review]
Thanks. Code looks good and seems to work fine.

One issue I noticed: If I import a new track that has a Serato beatgrid, i can see the beatgrid, but it's overwritten as soon as the analysis finished. There's no way to use the Serato beatgrid without resetting the Beatgrid afterwards and then reimporting the tags. I think that's not related to this branch, but I need to double check. Maybe it even affects 2.3.

**Comment by daschuer** [comment]
We have a preferences option for this.
By default it should not reanalyze. Maybe there is a bug if it is done anyway.

**Comment by uklotzde** [comment]
Behavior of the existing *import from source* didn't change and the new function is not used yet. It will be used soon by more complex refactoring of DlgTrackInfo which fixes the workflow to allow dependent/derived fields.

Related: https://bugs.launchpad.net/mixxx/+bug/1921646

The import/export and synchronization of beat grids is tricky and currently not really usable.

**Comment by Holzhaus** [comment]
The bug regarding Serato BeatGrids is unrelated and fixed in #3885.

@daschuer Do you want another look at this PR? Otherwise LGTM.

**Comment by uklotzde** [comment]
Found and fixed a bug that only affects main while testing https://bugs.launchpad.net/mixxx/+bug/1929311

**Comment by uklotzde** [comment]
Ping. I would like to move on now. This is blocking a follow-up PR.

File: src/track/track.cpp
Comment (daschuer) [inline src/track/track.cpp:42]: T&& is here a universal reference. It can be a lvalue or rvalue reference. So the T& version of this function is redundant.

File: src/track/trackrecord.cpp
Comment (daschuer) [inline src/track/trackrecord.cpp:116]: This std::move is redundant. We have a rvalue setter. Let's use it. 
```suggestion
        setMetadata(importedMetadata);
```

Comment (daschuer) [reply src/track/trackrecord.cpp:116]: This code is hard to read. Let's use setMetadata()

File: src/track/trackrecord.cpp
Comment (daschuer) [inline src/track/trackrecord.cpp:128]:
```suggestion
        setMetadataSynchronized(true);
```

**Review by daschuer — CHANGES_REQUESTED** [review]

File: src/track/track.cpp
Comment (uklotzde) [reply src/track/track.cpp:42]: :+1: After switching from std::move to std::forward this should work now. Didn't notice that the const-ref version became obsolete.

File: src/track/trackrecord.cpp
Comment (uklotzde) [reply src/track/trackrecord.cpp:116]: Good catch. setMetadata() also uses a universal reference and std::forward.

Comment (uklotzde) [reply src/track/trackrecord.cpp:116]: But we need to keep std::move() to ensure that we don't invoke an overload that accepts `const TrackMetadata&`. Currently there is none, but otherwise whenever it is added in the future this code would behave differently without std::move(). It is needed to express out intend.

Comment (uklotzde) [reply src/track/trackrecord.cpp:116]: See also: https://stackoverflow.com/a/32691106

File: src/track/track.cpp
Comment (uklotzde) [reply src/track/track.cpp:42]: Nope, that doesn't work. The const-ref overload is still needed.

File: src/track/trackrecord.cpp
Comment (daschuer) [reply src/track/trackrecord.cpp]: Ah OK, I did not know that.

File: src/track/track.cpp
Comment (daschuer) [reply src/track/track.cpp:42]: Ah OK, probably because of const.

**Review by daschuer — APPROVED** [review]
Thank you.

File: src/track/trackrecord.cpp
Comment (uklotzde) [reply src/track/trackrecord.cpp]: C++ has become too complex over time. Ideally the compiler would detect those pitfalls. I can't remember them all, too.

**Comment by daschuer** [comment]
Waiting for CI

**Pull request merged**
